numba/core/datamodel/models.py

```diff
@@ -103,11 +103,23 @@ def traverse_types(self):
         """
         Recursively list all frontend types involved in this model.
         """
-        return [self._fe_type] + self.inner_types()
+        types = [self._fe_type]
+        queue = [self]
+        while len(queue) > 0:
+            dm = queue[0]
+            queue.pop(0)
```

Comment:
`list.pop(0)` is O(n). Use `collections.deque` instead of `list` for `queue` to be more efficient.